File: 0653-Two Sum IV - Input is a BST/0653-Two Sum IV - Input is a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool findTarget(TreeNode* root, int k) {
        if (root == nullptr) {
            return false;
        }
        
        stack<TreeNode*> St;
        St.push(root);
        while (!St.empty()) {
            TreeNode* node = St.top();
            if (table.find(k - node->val) != table.end()) {
                return true;
            }
            table.insert(node->val);
            St.pop();
            
            if (node->right) {
                St.push(node->right);
            }
            
            if (node->left) {
                St.push(node->left);
            }
        }
        return false;
    }
    
private:
    unordered_set<int> table;
};

File: 0601-0700/0653-Two Sum IV - Input is a BST/0653-Two Sum IV - Input is a BST.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def findTarget(self, root: TreeNode, k: int) -> bool:
        if root is None:
            return False
        table = set()
        St = [root]
        while St:
            node = St.pop()
            if k - node.val in table:
                return True
            table.add(node.val)
            if node.right:
                St.append(node.right)
            if node.left:
                St.append(node.left)
        return False